api_v2/services/stream_service.go

```diff
@@ -2,59 +2,138 @@
+		return nil, err
+	} else if errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, e.WithStatus(http.StatusNotFound, errors.New("stream not found"))
 	}
 
-    return s, nil
+	return s, nil
 }
 
 // GetStreamsByCourseID retrieves all streams of a course by its id.
 func GetStreamsByCourseID(db *gorm.DB, courseID uint) ([]*model.Stream, error) {
-    var streams []*model.Stream
-    if err := db.Where("streams.course_id = ?", courseID).Find(&streams).Error; err != nil {
-        return nil, err
+	var streams []*model.Stream
+	if err := db.Where("streams.course_id = ?", courseID).Find(&streams).Error; err != nil {
+		return nil, err
 	}
 
-    return streams, nil
+	return streams, nil
 }
 
 func GetEnrolledOrPublicLiveStreams(db *gorm.DB, uID *uint) ([]*model.Stream, error) {
-    var streams []*model.Stream
-    if *uID == 0 {
-        err := db.Table("streams").
-        Joins("join courses on streams.course_id = courses.id").
-        Joins("left join course_users on courses.id = course_users.course_id").
-        Where("(course_users.user_id = ? OR courses.visibility = \"public\") AND streams.live_now = 1", *uID).
-        Find(&streams).Error
-        if err != nil {
-            return nil, err
-        }
-    } else {
-        err := db.Table("streams").
-        Select("DISTINCT streams.*").
-        Joins("join courses on streams.course_id = courses.id").
-        Joins("left join course_users on courses.id = course_users.course_id").
-        Where("(course_users.user_id = ? OR courses.visibility = \"public\" OR courses.visibility = \"loggedin\") AND streams.live_now = 1", *uID).
-        Find(&streams).Error
-        if err != nil {
-            return nil, err
-        }
-    }
-
-    
-    return streams, nil
-}
\ No newline at end of file
+	var streams []*model.Stream
+	if *uID == 0 {
+		err := db.Table("streams").
+			Joins("join courses on streams.course_id = courses.id").
+			Joins("left join course_users on courses.id = course_users.course_id").
+			Where("(course_users.user_id = ? OR courses.visibility = \"public\") AND streams.live_now = 1", *uID).
+			Find(&streams).Error
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		err := db.Table("streams").
+			Select("DISTINCT streams.*").
+			Joins("join courses on streams.course_id = courses.id").
+			Joins("left join course_users on courses.id = course_users.course_id").
+			Where("(course_users.user_id = ? OR courses.visibility = \"public\" OR courses.visibility = \"loggedin\") AND streams.live_now = 1", *uID).
+			Find(&streams).Error
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return streams, nil
+}
+
+// GetProgress retrieves the progress of a stream for a user.
+func GetProgress(db *gorm.DB, streamID uint, userID uint) (*model.StreamProgress, error) {
+	p := &model.StreamProgress{}
+
+	err := db.Where("stream_id = ? AND user_id = ?", streamID, userID).First(p).Error
+
+	if err != nil && !errors.Is(err, gorm.ErrRecordNotFound) {
+		return nil, err
```

Comment:
Return `nil, e.WithStatus(http.StatusInternalServerError, err)`. instead of `nil, err`